File: frontend/api/marketing.py
from __future__ import annotations

import asyncio
import base64
import html
import json
import re
from datetime import datetime, timezone
from typing import Any

from fastapi import Depends, FastAPI, HTTPException, Request
from fastapi.middleware.cors import CORSMiddleware

from api.backend.odoo_connector import OdooConnectorError, get_odoo_connector
from api.unified import hosted_get_current_user

MARKER = "ARAAK_MARKETING_V1:"
SOURCES = ["اعتماد", "فرصة", "منافس", "مناقصات", "إحالة مباشرة", "مصدر داخلي"]
MAX_FILES = 5
MAX_FILE_SIZE = 3 * 1024 * 1024

app = FastAPI(title="ARAAK Marketing Central Records Gateway")
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=False,
    allow_methods=["POST", "OPTIONS"],
    allow_headers=["*"],
)


def encode_metadata(metadata: dict[str, Any]) -> str:
    raw = json.dumps(metadata, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def decode_metadata(description: Any) -> dict[str, Any] | None:
    match = re.search(r"ARAAK_MARKETING_V1:([A-Za-z0-9_-]+)", str(description or ""))
    if not match:
        return None
    token = match.group(1) + "=" * (-len(match.group(1)) % 4)
    try:
        value = json.loads(base64.urlsafe_b64decode(token.encode("ascii")).decode("utf-8"))
        return value if isinstance(value, dict) else None
    except Exception:
        return None


def strip_html(value: Any) -> str:
    text = re.sub(r"<!--.*?-->", " ", str(value or ""), flags=re.S)
    text = re.sub(r"<[^>]+>", " ", text)
    return re.sub(r"\s+", " ", html.unescape(text)).strip()


def m2o_name(value: Any) -> str:
    return str(value[1] or "") if isinstance(value, (list, tuple)) and len(value) > 1 else ""


def created_id(value: Any) -> int | None:
    if isinstance(value, int):
        return value
    if isinstance(value, list) and value:
        if isinstance(value[0], int):
            return value[0]
        if isinstance(value[0], dict) and isinstance(value[0].get("id"), int):
            return value[0]["id"]
    if isinstance(value, dict) and isinstance(value.get("id"), int):
        return value["id"]
    return None


async def call(model: str, method: str, args: list[Any] | None = None, kwargs: dict[str, Any] | None = None) -> Any:
    connector = get_odoo_connector()
    return await asyncio.to_thread(
        connector._call_model_sync,
        model,
        method,
        args or [],
        kwargs or {},
    )


def description_html(metadata: dict[str, Any]) -> str:
    marker = f"<!--{MARKER}{encode_metadata(metadata)}-->"
    sections = []
    if metadata.get("description"):
        sections.append(f"<p><strong>الوصف:</strong> {html.escape(str(metadata['description']))}</p>")
    if metadata.get("requirements"):
        sections.append(f"<p><strong>المتطلبات:</strong> {html.escape(str(metadata['requirements']))}</p>")
    if metadata.get("reference"):
        sections.append(f"<p><strong>المرجع:</strong> {html.escape(str(metadata['reference']))}</p>")
    if metadata.get("source"):
        sections.append(f"<p><strong>المصدر:</strong> {html.escape(str(metadata['source']))}</p>")
    return marker + "\n".join(sections)


def attachment_view(row: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": int(row["id"]),
        "name": str(row.get("name") or "ملف"),
        "mime_type": str(row.get("mimetype") or "application/octet-stream"),
        "file_size": int(row.get("file_size") or 0),
        "created_at": row.get("create_date"),
    }


def record_view(row: dict[str, Any], attachments: dict[int, list[dict[str, Any]]]) -> dict[str, Any]:
    metadata = decode_metadata(row.get("description")) or {}
    row_id = int(row["id"])
    kind = "tender" if metadata.get("kind") == "tender" else "opportunity"
    return {
        "id": row_id,
        "kind": kind,
        "title": str(row.get("name") or ""),
        "reference": metadata.get("reference"),
        "client": row.get("partner_name") or metadata.get("client"),
        "entity": metadata.get("entity"),
        "city": row.get("city") or metadata.get("city"),
        "value": float(row.get("expected_revenue") or metadata.get("value") or 0) or None,
        "deadline": row.get("date_deadline") or metadata.get("deadline"),
        "publication_date": metadata.get("publication_date"),
        "description": metadata.get("description") or strip_html(row.get("description")),
        "requirements": metadata.get("requirements"),
        "source": metadata.get("source") or "مصدر داخلي",
        "source_url": metadata.get("source_url"),
        "status": metadata.get("status") or ("cancelled" if row.get("active") is False else "active"),
        "current_stage": metadata.get("current_stage") or m2o_name(row.get("stage_id")) or "الاستقبال",
        "stage_label": m2o_name(row.get("stage_id")) or metadata.get("current_stage") or "الاستقبال",
        "probability": float(row.get("probability") or 0),
        "owner": m2o_name(row.get("user_id")) or None,
        "team": m2o_name(row.get("team_id")) or None,
        "created_at": row.get("create_date"),
        "updated_at": row.get("write_date"),
        "attachments": attachments.get(row_id, []),
    }


async def list_records(kind: str | None = None) -> list[dict[str, Any]]:
    leads = await call(
        "crm.lead",
        "search_read",
        [[["description", "ilike", MARKER]]],
        {
            "fields": [
                "id", "name", "partner_name", "city", "expected_revenue", "date_deadline",
                "description", "stage_id", "probability", "active", "create_date", "write_date",
                "user_id", "team_id",
            ],
            "order": "create_date desc",
            "limit": 250,
        },
    )
    leads = leads if isinstance(leads, list) else []
    ids = [int(row["id"]) for row in leads if row.get("id")]
    grouped: dict[int, list[dict[str, Any]]] = {}
    if ids:
        rows = await call(
            "ir.attachment",
            "search_read",
            [[[
                "res_model", "=", "crm.lead"
            ], [
                "res_id", "in", ids
            ]]],
            {
                "fields": ["id", "name", "mimetype", "file_size", "create_date", "res_id"],
                "order": "create_date desc",
                "limit": 1000,
            },
        )
        for row in rows if isinstance(rows, list) else []:
            grouped.setdefault(int(row.get("res_id") or 0), []).append(attachment_view(row))
    records = [record_view(row, grouped) for row in leads]
    return [row for row in records if not kind or row["kind"] == kind]


async def create_record(payload: dict[str, Any], user: dict[str, Any]) -> dict[str, Any]:
    connector = get_odoo_connector()
    if connector.config.read_only:
        raise RuntimeError("التكامل المؤسسي مضبوط على القراءة فقط؛ يجب السماح بالكتابة في إعدادات ARAAK CEO.")

    kind = "tender" if payload.get("kind") == "tender" else "opportunity"
    record = payload.get("record") if isinstance(payload.get("record"), dict) else {}
    title = str(record.get("title") or "").strip()
    if not title:
        raise ValueError("عنوان الفرصة أو المنافسة مطلوب.")

    metadata = {
        "kind": kind,
        "reference": str(record.get("reference") or "").strip() or None,
        "client": str(record.get("client") or "").strip() or None,
        "entity": str(record.get("entity") or "").strip() or None,
        "city": str(record.get("city") or "").strip() or None,
        "value": float(record["value"]) if record.get("value") else None,
        "deadline": record.get("deadline") or None,
        "publication_date": record.get("publication_date") or None,
        "description": str(record.get("description") or "").strip() or None,
        "requirements": str(record.get("requirements") or "").strip() or None,
        "source": str(record.get("source") or "مصدر داخلي").strip(),
        "source_url": str(record.get("source_url") or "").strip() or None,
        "status": "in_progress" if kind == "tender" else "new",
        "current_stage": "الاستقبال" if kind == "tender" else "الدراسة الأولية",
        "created_by_email": user.get("email"),
        "created_by_role": user.get("role"),
        "created_at": datetime.now(timezone.utc).isoformat(),
    }
    values: dict[str, Any] = {
        "name": title,
        "type": "opportunity",
        "description": description_html(metadata),
    }
    optional = {
        "partner_name": str(record.get("client") or record.get("entity") or "").strip(),
        "city": str(record.get("city") or "").strip(),
        "expected_revenue": float(record["value"]) if record.get("value") else None,
        "date_deadline": record.get("deadline") or None,
    }
    values.update({key: value for key, value in optional.items() if value not in (None, "")})

    result = await call("crm.lead", "create", [[values]], {})
    lead_id = created_id(result)
    if not lead_id:
        raise RuntimeError("لم يُرجع السجل المركزي رقمًا صالحًا للسجل الجديد.")

    attachment_ids: list[int] = []
    files = payload.get("files") if isinstance(payload.get("files"), list) else []
    for item in files[:MAX_FILES]:
        if not isinstance(item, dict) or not item.get("name") or not item.get("data_base64"):
            continue
        if int(item.get("size") or 0) > MAX_FILE_SIZE:
            raise ValueError(f"الملف {item['name']} يتجاوز الحد المسموح.")
        attachment_values = {
            "name": str(item["name"]),
            "type": "binary",
            "datas": re.sub(r"^data:[^;]+;base64,", "", str(item["data_base64"])),
            "mimetype": str(item.get("mime_type") or "application/octet-stream"),
            "res_model": "crm.lead",
            "res_id": lead_id,
            "public": False,
        }
        attachment_result = await call("ir.attachment", "create", [[attachment_values]], {})
        attachment_id = created_id(attachment_result)
        if attachment_id:
            attachment_ids.append(attachment_id)

    records = await list_records(kind)
    return {
        "record": next((row for row in records if row["id"] == lead_id), {"id": lead_id, "kind": kind, "title": title}),
        "attachment_ids": attachment_ids,
    }


async def download_attachment(attachment_id: Any) -> dict[str, Any]:
    rows = await call(
        "ir.attachment",
        "read",
        [[int(attachment_id)]],
        {"fields": ["id", "name", "mimetype", "datas", "file_size", "res_model", "res_id"], "load": None},
    )
    item = rows[0] if isinstance(rows, list) and rows else None
    if not item or item.get("res_model") != "crm.lead":
        raise RuntimeError("الملف المطلوب غير متاح.")
    leads = await call("crm.lead", "read", [[int(item["res_id"])]], {"fields": ["id", "description"], "load": None})
    lead = leads[0] if isinstance(leads, list) and leads else None
    if not lead or not decode_metadata(lead.get("description")):
        raise RuntimeError("الملف غير مرتبط بسجل مسموح.")
    return {
        "id": int(item["id"]),
        "name": str(item.get("name") or "file"),
        "mime_type": str(item.get("mimetype") or "application/octet-stream"),
        "file_size": int(item.get("file_size") or 0),
        "data_base64": str(item.get("datas") or ""),
    }


async def execute(payload: dict[str, Any], user: dict[str, Any]) -> dict[str, Any]:
    action = str(payload.get("action") or "list")
    if action == "sources":
        return {"ok": True, "sources": SOURCES}
    if action == "status":
        connector = get_odoo_connector()
        state = await connector.status(check=True)
        return {
            "ok": True,
            "configured": bool(connector.config.configured),
            "connected": bool(state.get("connected")),
            "read_only": bool(connector.config.read_only),
            "message": state.get("message"),
        }
    if action == "list":
        kind = payload.get("kind") if payload.get("kind") in ("opportunity", "tender") else None
        records = await list_records(kind)
        return {"ok": True, "records": records, "total": len(records)}
    if action == "create":
        return {"ok": True, **(await create_record(payload, user))}
    if action == "download":
        return {"ok": True, "file": await download_attachment(payload.get("attachment_id"))}
    raise ValueError("العملية المطلوبة غير مدعومة.")


@app.post("/")
@app.post("/api/marketing")
async def marketing_gateway(request: Request, user=Depends(hosted_get_current_user)):
    try:
        payload = await request.json()
    except Exception:
        payload = {}
    try:
        return await execute(payload, user)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    except OdooConnectorError as error:
        raise HTTPException(status_code=503, detail=str(error)) from error
    except RuntimeError as error:
        raise HTTPException(status_code=502, detail=str(error)) from error
    except Exception as error:
        raise HTTPException(status_code=500, detail="تعذر تنفيذ العملية في السجل المركزي.") from error

File: api/marketing.js
import crypto from "node:crypto";

const JWT_SECRET = process.env.JWT_SECRET || "nexgen-vercel-hosted-demo-secret-2026";
const MARKER = "ARAAK_MARKETING_V1:";
const DEFAULT_SOURCES = ["اعتماد", "فرصة", "منافس", "مناقصات", "إحالة مباشرة", "مصدر داخلي"];
const MAX_FILE_SIZE = 3 * 1024 * 1024;
const MAX_FILES = 5;

function setHeaders(response) {
  response.setHeader("Cache-Control", "no-store");
  response.setHeader("Content-Type", "application/json; charset=utf-8");
  response.setHeader("Access-Control-Allow-Origin", "*");
  response.setHeader("Access-Control-Allow-Headers", "Authorization, Content-Type");
  response.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
}

function send(response, status, payload) {
  setHeaders(response);
  return response.status(status).json(payload);
}

function bodyOf(request) {
  if (request.body == null) return {};
  if (typeof request.body === "object") return request.body;
  try {
    return JSON.parse(request.body);
  } catch {
    return {};
  }
}

function parseInstitutionalToken(request) {
  const authorization = request.headers.authorization || request.headers.Authorization || "";
  const token = String(authorization).replace(/^Bearer\s+/i, "").trim();
  if (!token) return null;

  try {
    const [payloadPart, signaturePart] = token.split(".");
    if (!payloadPart || !signaturePart) return null;
    const expected = crypto.createHmac("sha256", JWT_SECRET).update(payloadPart).digest("base64url");
    if (expected.length !== signaturePart.length) return null;
    if (!crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(signaturePart))) return null;
    const payload = JSON.parse(Buffer.from(payloadPart, "base64url").toString("utf8"));
    if (!payload?.email || Number(payload.exp || 0) < Date.now()) return null;
    return payload;
  } catch {
    return null;
  }
}

function odooConfig() {
  return {
    enabled: String(process.env.ODOO_ENABLED || "true").toLowerCase() !== "false",
    url: String(process.env.ODOO_URL || "https://araakceo.odoo.com").replace(/\/+$/, ""),
    database: String(process.env.ODOO_DATABASE || "araakceo").trim(),
    apiKey: String(process.env.ODOO_API_KEY || "").trim(),
    language: String(process.env.ODOO_LANGUAGE || "en_US").trim(),
    timeoutMs: Math.max(3000, Number(process.env.ODOO_TIMEOUT_MS || 20000)),
    readOnly: String(process.env.ODOO_READ_ONLY || "false").toLowerCase() === "true",
  };
}

async function odooCall(model, method, parameters = {}) {
  const config = odooConfig();
  if (!config.enabled) throw new Error("التكامل المؤسسي غير مفعل.");
  if (!config.apiKey) throw new Error("مفتاح التكامل المؤسسي غير مضاف إلى بيئة ARAAK CEO.");

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), config.timeoutMs);
  try {
    const response = await fetch(`${config.url}/json/2/${encodeURIComponent(model)}/${encodeURIComponent(method)}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=utf-8",
        Authorization: `bearer ${config.apiKey}`,
        "X-Odoo-Database": config.database,
        "User-Agent": "ARAAK-CEO-Marketing-Gateway/1.0",
      },
      body: JSON.stringify({
        context: { lang: config.language },
        ...parameters,
      }),
      signal: controller.signal,
    });

    const payload = await response.json().catch(() => null);
    if (!response.ok) {
      const message = payload?.message || payload?.name || payload?.error || `HTTP ${response.status}`;
      throw new Error(`تعذر تنفيذ العملية في السجل المركزي: ${message}`);
    }
    return payload;
  } finally {
    clearTimeout(timeout);
  }
}

function compact(object) {
  return Object.fromEntries(Object.entries(object).filter(([, value]) => value !== undefined && value !== null && value !== ""));
}

function escapeHtml(value) {
  return String(value || "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;");
}

function stripHtml(value) {
  return String(value || "")
    .replace(/<!--.*?-->/gs, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/\s+/g, " ")
    .trim();
}

function encodeMetadata(metadata) {
  return Buffer.from(JSON.stringify(metadata), "utf8").toString("base64url");
}

function decodeMetadata(description) {
  const match = String(description || "").match(/ARAAK_MARKETING_V1:([A-Za-z0-9_-]+)/);
  if (!match) return null;
  try {
    return JSON.parse(Buffer.from(match[1], "base64url").toString("utf8"));
  } catch {
    return null;
  }
}

function buildDescription(metadata, description, requirements) {
  const marker = `<!--${MARKER}${encodeMetadata(metadata)}-->`;
  const sections = [
    description ? `<p><strong>الوصف:</strong> ${escapeHtml(description)}</p>` : "",
    requirements ? `<p><strong>المتطلبات:</strong> ${escapeHtml(requirements)}</p>` : "",
    metadata.reference ? `<p><strong>المرجع:</strong> ${escapeHtml(metadata.reference)}</p>` : "",
    metadata.source ? `<p><strong>المصدر:</strong> ${escapeHtml(metadata.source)}</p>` : "",
  ].filter(Boolean);
  return `${marker}${sections.join("\n")}`;
}

function createdId(result) {
  if (typeof result === "number") return result;
  if (Array.isArray(result)) {
    const first = result[0];
    if (typeof first === "number") return first;
    if (first && typeof first.id === "number") return first.id;
  }
  if (result && typeof result.id === "number") return result.id;
  return null;
}

function many2oneLabel(value) {
  return Array.isArray(value) ? String(value[1] || "") : "";
}

function normaliseAttachment(record) {
  return {
    id: Number(record.id),
    name: String(record.name || "ملف"),
    mime_type: String(record.mimetype || "application/octet-stream"),
    file_size: Number(record.file_size || 0),
    created_at: record.create_date || null,
  };
}

function normaliseLead(record, attachmentsByLead) {
  const metadata = decodeMetadata(record.description) || {};
  return {
    id: Number(record.id),
    kind: metadata.kind === "tender" ? "tender" : "opportunity",
    title: String(record.name || ""),
    reference: metadata.reference || null,
    client: record.partner_name || metadata.client || null,
    entity: metadata.entity || null,
    city: record.city || metadata.city || null,
    value: Number(record.expected_revenue || metadata.value || 0) || null,
    deadline: record.date_deadline || metadata.deadline || null,
    publication_date: metadata.publication_date || null,
    description: metadata.description || stripHtml(record.description),
    requirements: metadata.requirements || null,
    source: metadata.source || "مصدر داخلي",
    source_url: metadata.source_url || null,
    status: metadata.status || (record.active === false ? "cancelled" : "active"),
    current_stage: metadata.current_stage || many2oneLabel(record.stage_id) || "الاستقبال",
    stage_label: many2oneLabel(record.stage_id) || metadata.current_stage || "الاستقبال",
    probability: Number(record.probability || 0),
    owner: many2oneLabel(record.user_id) || null,
    team: many2oneLabel(record.team_id) || null,
    created_at: record.create_date || null,
    updated_at: record.write_date || null,
    attachments: attachmentsByLead.get(Number(record.id)) || [],
  };
}

async function listRecords(kind) {
  const records = await odooCall("crm.lead", "search_read", {
    domain: [["description", "ilike", MARKER]],
    fields: [
      "id", "name", "partner_name", "city", "expected_revenue", "date_deadline",
      "description", "stage_id", "probability", "active", "create_date", "write_date",
      "user_id", "team_id",
    ],
    order: "create_date desc",
    limit: 250,
  });

  const leads = Array.isArray(records) ? records : [];
  const ids = leads.map((record) => Number(record.id)).filter(Boolean);
  const attachmentsByLead = new Map();

  if (ids.length > 0) {
    const attachments = await odooCall("ir.attachment", "search_read", {
      domain: [["res_model", "=", "crm.lead"], ["res_id", "in", ids]],
      fields: ["id", "name", "mimetype", "file_size", "create_date", "res_id"],
      order: "create_date desc",
      limit: 1000,
    });
    for (const attachment of Array.isArray(attachments) ? attachments : []) {
      const leadId = Number(attachment.res_id);
      const current = attachmentsByLead.get(leadId) || [];
      current.push(normaliseAttachment(attachment));
      attachmentsByLead.set(leadId, current);
    }
  }

  return leads
    .map((record) => normaliseLead(record, attachmentsByLead))
    .filter((record) => !kind || record.kind === kind);
}

async function createRecord(payload, user) {
  const config = odooConfig();
  if (config.readOnly) throw new Error("التكامل المؤسسي مضبوط حاليًا على القراءة فقط.");

  const kind = payload.kind === "tender" ? "tender" : "opportunity";
  const record = payload.record || {};
  if (!String(record.title || "").trim()) throw new Error("عنوان الفرصة أو المنافسة مطلوب.");

  const metadata = {
    kind,
    reference: String(record.reference || "").trim() || null,
    client: String(record.client || "").trim() || null,
    entity: String(record.entity || "").trim() || null,
    city: String(record.city || "").trim() || null,
    value: record.value ? Number(record.value) : null,
    deadline: record.deadline || null,
    publication_date: record.publication_date || null,
    description: String(record.description || "").trim() || null,
    requirements: String(record.requirements || "").trim() || null,
    source: String(record.source || "مصدر داخلي").trim(),
    source_url: String(record.source_url || "").trim() || null,
    status: kind === "tender" ? "in_progress" : "new",
    current_stage: kind === "tender" ? "الاستقبال" : "الدراسة الأولية",
    created_by_email: user.email,
    created_by_role: user.role || null,
    created_at: new Date().toISOString(),
  };

  const values = compact({
    name: String(record.title).trim(),
    type: "opportunity",
    partner_name: String(record.client || record.entity || "").trim() || undefined,
    city: String(record.city || "").trim() || undefined,
    expected_revenue: record.value ? Number(record.value) : undefined,
    date_deadline: record.deadline || undefined,
    description: buildDescription(metadata, record.description, record.requirements),
  });

  const result = await odooCall("crm.lead", "create", { vals_list: [values] });
  const leadId = createdId(result);
  if (!leadId) throw new Error("تم إرسال السجل، لكن لم يُرجع النظام رقمًا مرجعيًا صالحًا.");

  const files = Array.isArray(payload.files) ? payload.files.slice(0, MAX_FILES) : [];
  const createdAttachments = [];
  for (const file of files) {
    const size = Number(file.size || 0);
    if (!file.name || !file.data_base64) continue;
    if (size > MAX_FILE_SIZE) throw new Error(`الملف ${file.name} يتجاوز الحد المسموح عبر البوابة.`);

    const attachmentValues = {
      name: String(file.name),
      type: "binary",
      datas: String(file.data_base64).replace(/^data:[^;]+;base64,/, ""),
      mimetype: String(file.mime_type || "application/octet-stream"),
      res_model: "crm.lead",
      res_id: leadId,
      public: false,
    };
    const attachmentResult = await odooCall("ir.attachment", "create", { vals_list: [attachmentValues] });
    const attachmentId = createdId(attachmentResult);
    if (attachmentId) createdAttachments.push(attachmentId);
  }

  const records = await listRecords(kind);
  return {
    record: records.find((item) => item.id === leadId) || { id: leadId, kind, title: values.name },
    attachment_ids: createdAttachments,
  };
}

async function downloadAttachment(attachmentId) {
  const attachmentRows = await odooCall("ir.attachment", "read", {
    ids: [Number(attachmentId)],
    fields: ["id", "name", "mimetype", "datas", "file_size", "res_model", "res_id"],
    load: null,
  });
  const attachment = Array.isArray(attachmentRows) ? attachmentRows[0] : null;
  if (!attachment || attachment.res_model !== "crm.lead") throw new Error("الملف المطلوب غير متاح.");

  const leadRows = await odooCall("crm.lead", "read", {
    ids: [Number(attachment.res_id)],
    fields: ["id", "description"],
    load: null,
  });
  const lead = Array.isArray(leadRows) ? leadRows[0] : null;
  if (!lead || !decodeMetadata(lead.description)) throw new Error("الملف غير مرتبط بسجل مسموح.");

  return {
    id: Number(attachment.id),
    name: String(attachment.name || "file"),
    mime_type: String(attachment.mimetype || "application/octet-stream"),
    file_size: Number(attachment.file_size || 0),
    data_base64: String(attachment.datas || ""),
  };
}

export default async function handler(request, response) {
  setHeaders(response);
  if (request.method === "OPTIONS") return response.status(204).end();
  if (request.method !== "POST") return send(response, 405, { ok: false, message: "Method not allowed" });

  const user = parseInstitutionalToken(request);
  if (!user) return send(response, 401, { ok: false, message: "انتهت الجلسة المؤسسية؛ سجّل الدخول من جديد." });

  const payload = bodyOf(request);
  const action = String(payload.action || "list");

  try {
    if (action === "sources") return send(response, 200, { ok: true, sources: DEFAULT_SOURCES });
    if (action === "status") {
      const config = odooConfig();
      return send(response, 200, {
        ok: true,
        configured: config.enabled && Boolean(config.apiKey),
        read_only: config.readOnly,
      });
    }
    if (action === "list") {
      const records = await listRecords(payload.kind === "tender" ? "tender" : payload.kind === "opportunity" ? "opportunity" : null);
      return send(response, 200, { ok: true, records, total: records.length });
    }
    if (action === "create") {
      const result = await createRecord(payload, user);
      return send(response, 201, { ok: true, ...result });
    }
    if (action === "download") {
      const file = await downloadAttachment(payload.attachment_id);
      return send(response, 200, { ok: true, file });
    }
    return send(response, 400, { ok: false, message: "العملية المطلوبة غير مدعومة." });
  } catch (error) {
    console.error("marketing gateway failed", error);
    return send(response, 502, {
      ok: false,
      message: error instanceof Error ? error.message : "تعذر تنفيذ العملية في السجل المركزي.",
    });
  }
}
